### ISSUE TITLE
Preliminary json schema and discussion

### ISSUE THREAD

**Issue by drewhutchison** [Denver-Devs/rolodex#2]
Continuing the discussion from #1, the following is my first shot at a json schema matching @aaronagray's sample data:

```
{
  "type": "array",
  "items": {
    "type": "object",
    "properties": {
      "picture": {"type": "string"},
      "website": {"type": "string"},
      "name": {
        "type": "object",
        "properties": {
          "first": {"type": "string"},
          "last": {"type": "string"}
        },
        "required": [
          "first",
          "last"
        ],
        "additionalProperties": false
      },
      "company": {"type": "string"},
      "email": {"type": "string"},
      "pronouns": {
        "type": "object",
        "properties": {
          "personal": {"type": "string"},
          "object": {"type": "string"},
          "reflexive": {"type": "string"},
          "possessive": {"type": "string"}
        },
        "additionalProperties": false
      },
      "twitter": {"type": "string"},
      "phone": {"type": "string"},
      "about": {"type": "string"},
      "city": {"type": "string"},
      "state": {"type": "string"},
      "skills": {
        "type": "array",
        "items": {"type": "string"}
      }
    },
    "required": [
      "name"
    ],
    "additionalProperties": false
  }
}
```

I propose we include this as, e.g., `schema.json`. 

Then, (with [ajv-cli](https://github.com/jessedc/ajv-cli) installed), `ajv -s schema.json -d members.json` will validate the member roll. I'm not married to `ajv` as a json-schema implementation, but the CLI tool obviates the need for full-on CI or any support packages or executables.

### Open questions

1. This whole project came up because someone was curious about everyone's github ID. We should therefore probably include that somewhere

2. For that matter, we should also include everyone's slack username

3. Also gitlab, SO, perlmonks, etc

4. More importantly, we should lock down a list of required fields. I propose minimally `name` and slack username, with the rest being optional

5. It might be a little early, but json-schema also has pretty decent string validation support, so we might discuss that in regard to URLs, email, phone no, state, etc.

6. As @tostal mentioned on #1, there is an argument to be made against this being an array structure at the top level.

**Comment by LegitTalon** [comment]
```
links: {
  <nickname>: <url>
}
```

**Comment by drewhutchison** [comment]
[gist incorporating Talon's suggestion](https://gist.github.com/drewhutchison/df4f61322d22970c76d2/revisions)

**Comment by aaronagray** [comment]
Nice. The schema looks great. I agree with you thoughts on the open questions and your suggested solutions. I've never used ajv-cli, but it seems like a solid immediate solution to me.

**Comment by killtheliterate** [comment]
SteamID, etc... so we can waste time together!

**Comment by drewhutchison** [comment]
@aaronagray yeah, I've used probably about five js json-schema implementations over the years and found something to hate about each of them. Sold on this one having the cli just so we can keep it to three files and none of those is a `package.json`.... yet.

@killtheliterate specifying `additionalProperties` like that will validate arbitrary keys in the `links` section as long as they're string-valued :)

**Comment by sreynen** [comment]
I don't work for a company, and have met some other denver devs with similar non-company employment. I suggest that be "employer" instead.

More generally, I vote for a pave-the-cowpaths approach to this. Do a minimal schema (maybe just the required fields), see what kind of info people submit beyond that, and refine the schema based on real data.

**Comment by DrewDahlman** [comment]
Updated [Gist](https://gist.github.com/DrewDahlman/cefce3214d72363216ff) based on feedback from @sreynen

**Comment by DrewDahlman** [comment]
@drewhutchison and I were talking earlier today about getting people from DD to use this. We have the option to scrape the users list pretty easily via the slack API to setup initial things, but it's going to become a total nightmare to manage so there could be validity to having a form OR it spits out a file for each person.

I don't want to suddenly put EVERYONE on a list somewhere and be like "hey you didn't do anything but you're listed here now, manage yourself! k thx bye." so we should put some thought into how we get the community into this without being too heavy handed. I will start a new issue with this for us to talk about.
